File: playground/backend/reasoning_overview.py
"""Reasoning Overview ViewModel builder for Phase 8D."""

from __future__ import annotations

import copy
import json
from typing import Any

CONTRACT_SCHEMA = "reasonscript-playground-reasoning-overview/1.0"
RUNTIME_SCHEMA = "reasonscript-reasoning-runtime-prototype/1.0"
MODEL_SCHEMA = "reasonscript-reasoning-model/1.0"
EVALUATION_SCHEMA = "reasonscript-reasoning-evaluation-report/1.0"

SEVERITY_ORDER = {"fatal": 0, "error": 1, "warning": 2, "info": 3}


def build_reasoning_overview_view_model(runtime_result: dict[str, Any] | None) -> dict[str, Any]:
    runtime = runtime_result if isinstance(runtime_result, dict) else {}
    model = _as_dict(runtime.get("reasoning_model"))
    report = _as_dict(runtime.get("evaluation_report"))
    diagnostics = _diagnostics(runtime, model, report)

    return _stable({
        "schemaVersion": CONTRACT_SCHEMA,
        "sourceRef": _source_ref(runtime),
        "runtimeStatus": _runtime_status(runtime, model, report, diagnostics),
        "modelSummary": _model_summary(model),
        "pipelineStatus": _pipeline_status(runtime),
        "inputState": _input_state(model),
        "reasoningPath": _reasoning_path(model),
        "knowledgeEmission": _knowledge_emission(model),
        "evaluationReport": _evaluation_report(report),
        "diagnostics": _diagnostics_summary(diagnostics),
        "rawArtifacts": {
            "runtimeResult": copy.deepcopy(runtime),
            "reasoningModel": copy.deepcopy(model),
            "evaluationReport": copy.deepcopy(report),
        },
    })


def serialize_reasoning_overview_view_model(view_model: dict[str, Any]) -> str:
    return json.dumps(_stable(view_model), ensure_ascii=False, indent=2, sort_keys=True, default=str) + "\n"


def _source_ref(runtime: dict[str, Any]) -> dict[str, Any]:
    source_ref = _as_dict(runtime.get("source_ref"))
    return {
        "sourceId": str(source_ref.get("source_id") or "unavailable"),
        "sourceKind": str(source_ref.get("source_kind") or "unknown"),
        **({"sourcePath": str(source_ref["source_path"])} if source_ref.get("source_path") else {}),
    }


def _runtime_status(
    runtime: dict[str, Any],
    model: dict[str, Any],
    report: dict[str, Any],
    diagnostics: list[dict[str, Any]],
) -> dict[str, Any]:
    pipeline = _as_dict(runtime.get("pipeline_status"))
    fatal = [item for item in diagnostics if item["severity"] == "fatal"]
    return {
        "status": _runtime_status_value(pipeline.get("status")),
        "runId": str(runtime.get("run_id") or ""),
        "hasReasoningModel": bool(model),
        "hasEvaluationReport": bool(report),
        "diagnosticCount": len(diagnostics),
        "fatalDiagnosticCount": len(fatal),
    }


def _model_summary(model: dict[str, Any]) -> dict[str, Any]:
    input_state = _as_dict(model.get("input_state"))
    paths = _as_list(model.get("reasoning_paths"))
    steps = [step for path in paths if isinstance(path, dict) for step in _as_list(path.get("steps")) if isinstance(step, dict)]
    target = _as_dict(model.get("evaluation_target"))
    return {
        "available": bool(model),
        "modelId": str(model.get("model_id") or "ReasoningModel unavailable"),
        "modelSchemaVersion": str(model.get("schema_version") or ""),
        "selectedPathId": str(model.get("selected_path_id") or ""),
        "inputUnitCount": len(_as_list(input_state.get("units"))),
        "inputRelationCount": len(_as_list(input_state.get("relations"))),
        "reasoningPathCount": len(paths),
        "reasoningStepCount": len(steps),
        "knowledgeEmissionCount": len(_as_list(model.get("knowledge_emissions"))),
        "evaluationGoal": str(target.get("goal") or ""),
        "requiredChecks": [str(item) for item in _as_list(target.get("required_checks"))],
    }


def _pipeline_status(runtime: dict[str, Any]) -> dict[str, Any]:
    status = _as_dict(runtime.get("pipeline_status"))
    return {
        "status": _runtime_status_value(status.get("status")),
        "parserPassed": bool(status.get("parser_passed")),
        "reasonIrAvailable": bool(status.get("reason_ir_available")),
        "executionPlanAvailable": bool(status.get("execution_plan_available")),
        "simulationAvailable": bool(status.get("simulation_available")),
        "knowledgeAvailable": bool(status.get("knowledge_available")),
        "diagnosticsCount": int(status.get("diagnostics_count") or 0),
    }


def _input_state(model: dict[str, Any]) -> dict[str, Any]:
    input_state = _as_dict(model.get("input_state"))
    return {
        "inputId": str(input_state.get("input_id") or ""),
        "inputKind": str(input_state.get("input_kind") or ""),
        "units": [
            {
                "unitId": str(unit.get("unit_id") or ""),
                "unitType": str(unit.get("unit_type") or ""),
                "value": unit.get("value"),
            }
            for unit in _as_list(input_state.get("units"))
            if isinstance(unit, dict)
        ],
        "relations": [
            {
                "relationId": str(relation.get("relation_id") or ""),
                "relationType": str(relation.get("relation_type") or ""),
                "source": str(relation.get("source") or ""),
                "target": str(relation.get("target") or ""),
            }
            for relation in _as_list(input_state.get("relations"))
            if isinstance(relation, dict)
        ],
    }


def _reasoning_path(model: dict[str, Any]) -> dict[str, Any]:
    selected_path_id = str(model.get("selected_path_id") or "")
    paths = [_path_item(path) for path in _as_list(model.get("reasoning_paths")) if isinstance(path, dict)]
    selected = next((path for path in paths if path["pathId"] == selected_path_id), {})
    return {
        "selectedPathId": selected_path_id,
        "selectedPathSignature": str(selected.get("pathSignature") or ""),
        "paths": paths,
    }


def _path_item(path: dict[str, Any]) -> dict[str, Any]:
    return {
        "pathId": str(path.get("path_id") or ""),
        "pathSignature": str(path.get("path_signature") or ""),
        "status": _path_status(path.get("status")),
        "steps": [
            {
                "stepId": str(step.get("step_id") or ""),
                "stepType": str(step.get("step_type") or ""),
                "source": str(step.get("source") or ""),
                "operation": str(step.get("operation") or ""),
                "target": str(step.get("target") or ""),
                "evidenceRefs": [str(ref) for ref in _as_list(step.get("evidence_refs"))],
            }
            for step in _as_list(path.get("steps"))
            if isinstance(step, dict)
        ],
    }


def _knowledge_emission(model: dict[str, Any]) -> dict[str, Any]:
    emissions = []
    for item in _as_list(model.get("knowledge_emissions")):
        if not isinstance(item, dict):
            continue
        emissions.append({
            "knowledgeId": str(item.get("knowledge_id") or ""),
            "sourceStepId": str(item.get("source_step_id") or ""),
            "relation": str(item.get("relation") or ""),
            "source": str(item.get("source") or ""),
            "target": str(item.get("target") or ""),
            "evidencePath": [str(ref) for ref in _as_list(item.get("evidence_path"))],
            "pathSignature": str(item.get("path_signature") or ""),
        })
    return {"emissions": emissions}


def _evaluation_report(report: dict[str, Any]) -> dict[str, Any]:
    summary = _as_dict(report.get("summary"))
    return {
        "available": bool(report),
        "reportId": str(report.get("report_id") or "EvaluationReport unavailable"),
        "status": _eval_status(summary.get("status")),
        "passed": bool(summary.get("passed")),
        "requiredChecksPassed": bool(summary.get("required_checks_passed")),
        "totalChecks": int(summary.get("total_checks") or 0),
        "passedChecks": int(summary.get("passed_checks") or 0),
        "failedChecks": int(summary.get("failed_checks") or 0),
        "warningChecks": int(summary.get("warning_checks") or 0),
        "fatalDiagnostics": int(summary.get("fatal_diagnostics") or 0),
        "checks": [
            {
                "checkId": str(check.get("check_id") or ""),
                "checkType": str(check.get("check_type") or ""),
                "required": bool(check.get("required")),
                "status": _check_status(check.get("status")),
                "passed": bool(check.get("passed")),
                "message": str(check.get("message") or ""),
                "evidenceRefs": [str(ref) for ref in _as_list(check.get("evidence_refs"))],
                "details": _as_dict(check.get("details")),
            }
            for check in _as_list(report.get("checks"))
            if isinstance(check, dict)
        ],
    }


def _diagnostics(runtime: dict[str, Any], model: dict[str, Any], report: dict[str, Any]) -> list[dict[str, Any]]:
    items = (
        _diagnostic_items("runtime", runtime.get("diagnostics"))
        + _diagnostic_items("reasoning_model", model.get("diagnostics"))
        + _diagnostic_items("evaluation_report", report.get("diagnostics"))
    )
    if runtime and runtime.get("schema_version") != RUNTIME_SCHEMA:
        items.append(_diag("runtime", "RO-ART-001", "error", "Unsupported ReasoningRuntimeResult schema version.", "schema_version"))
    if model and model.get("schema_version") != MODEL_SCHEMA:
        items.append(_diag("reasoning_model", "RO-ART-002", "error", "Unsupported ReasoningModel schema version.", "schema_version"))
    if report and report.get("schema_version") != EVALUATION_SCHEMA:
        items.append(_diag("evaluation_report", "RO-ART-003", "error", "Unsupported ReasoningEvaluationReport schema version.", "schema_version"))
    return sorted(items, key=lambda item: (SEVERITY_ORDER[item["severity"]], item["source"], item["code"], item.get("location", ""), item["message"]))


def _diagnostics_summary(items: list[dict[str, Any]]) -> dict[str, Any]:
    return {
        "total": len(items),
        "fatal": sum(1 for item in items if item["severity"] == "fatal"),
        "error": sum(1 for item in items if item["severity"] == "error"),
        "warning": sum(1 for item in items if item["severity"] == "warning"),
        "info": sum(1 for item in items if item["severity"] == "info"),
        "items": items,
    }


def _diagnostic_items(source: str, value: Any) -> list[dict[str, Any]]:
    result = []
    for item in _as_list(value):
        if not isinstance(item, dict):
            continue
        result.append(_diag(
            source,
            str(item.get("code") or "RO-DIAG-UNKNOWN"),
            _severity(item.get("severity")),
            str(item.get("message") or ""),
            str(item["location"]) if item.get("location") is not None else None,
        ))
    return result


def _diag(source: str, code: str, severity: str, message: str, location: str | None) -> dict[str, Any]:
    value = {"source": source, "code": code, "severity": severity, "message": message}
    if location:
        value["location"] = location
    return value


def _runtime_status_value(value: Any) -> str:
    return str(value) if value in {"passed", "failed", "partial", "fatal"} else "unknown"


def _path_status(value: Any) -> str:
    return str(value) if value in {"selected", "candidate", "rejected", "failed"} else "unknown"


def _eval_status(value: Any) -> str:
    return str(value) if value in {"passed", "failed", "warning", "fatal"} else "unknown"


def _check_status(value: Any) -> str:
    return str(value) if value in {"passed", "failed", "warning", "skipped", "fatal"} else "unknown"


def _severity(value: Any) -> str:
    text = str(value or "info").lower()
    return text if text in SEVERITY_ORDER else "info"


def _stable(value: Any) -> Any:
    if isinstance(value, list):
        return [_stable(item) for item in value]
    if isinstance(value, dict):
        return {key: _stable(value[key]) for key in sorted(value)}
    return value


def _as_dict(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def _as_list(value: Any) -> list[Any]:
    return value if isinstance(value, list) else []

File: apps/reasonscript-ide/ui/src/viewModels/reasoningOverview.ts
import type { ProjectState } from "../types";
import type {
  ReasoningDiagnosticItemViewModel,
  ReasoningRuntimeViewModel,
} from "../types/reasoningOverview";

const SCHEMA = "reasonscript-playground-reasoning-overview/1.0" as const;
const SEVERITY_ORDER: Record<string, number> = { fatal: 0, error: 1, warning: 2, info: 3 };

const unavailable: ReasoningRuntimeViewModel = {
  schemaVersion: SCHEMA,
  sourceRef: { sourceId: "unavailable", sourceKind: "unknown" },
  runtimeStatus: {
    status: "unknown",
    runId: "",
    hasReasoningModel: false,
    hasEvaluationReport: false,
    diagnosticCount: 0,
    fatalDiagnosticCount: 0,
  },
  modelSummary: {
    available: false,
    modelId: "ReasoningModel unavailable",
    modelSchemaVersion: "",
    selectedPathId: "",
    inputUnitCount: 0,
    inputRelationCount: 0,
    reasoningPathCount: 0,
    reasoningStepCount: 0,
    knowledgeEmissionCount: 0,
    evaluationGoal: "",
    requiredChecks: [],
  },
  pipelineStatus: {
    status: "unknown",
    parserPassed: false,
    reasonIrAvailable: false,
    executionPlanAvailable: false,
    simulationAvailable: false,
    knowledgeAvailable: false,
    diagnosticsCount: 0,
  },
  inputState: { inputId: "", inputKind: "", units: [], relations: [] },
  reasoningPath: { selectedPathId: "", selectedPathSignature: "", paths: [] },
  knowledgeEmission: { emissions: [] },
  evaluationReport: {
    available: false,
    reportId: "EvaluationReport unavailable",
    status: "unknown",
    passed: false,
    requiredChecksPassed: false,
    totalChecks: 0,
    passedChecks: 0,
    failedChecks: 0,
    warningChecks: 0,
    fatalDiagnostics: 0,
    checks: [],
  },
  diagnostics: { total: 0, fatal: 0, error: 0, warning: 0, info: 0, items: [] },
  rawArtifacts: {},
};

export function buildReasoningOverviewViewModel(projectState: ProjectState | null): ReasoningRuntimeViewModel {
  const backendVm = asRecord(projectState?.reasoning_overview);
  if (backendVm?.schemaVersion === SCHEMA) {
    const vm = backendVm as unknown as ReasoningRuntimeViewModel;
    return {
      ...vm,
      rawArtifacts: {
        ...(vm.rawArtifacts ?? {}),
        viewModel: vm,
      },
    };
  }

  const runtime = asRecord(projectState?.reasoning_runtime);
  if (!runtime) {
    return {
      ...unavailable,
      rawArtifacts: {
        runtimeResult: projectState?.reasoning_runtime as Record<string, unknown> | undefined,
        reasoningModel: projectState?.reasoning_model as Record<string, unknown> | undefined,
        evaluationReport: projectState?.reasoning_evaluation_report as Record<string, unknown> | undefined,
      },
    };
  }

  const model = asRecord(runtime.reasoning_model) ?? asRecord(projectState?.reasoning_model) ?? {};
  const report = asRecord(runtime.evaluation_report) ?? asRecord(projectState?.reasoning_evaluation_report) ?? {};
  const diagnostics = collectDiagnostics(runtime, model, report);
  const input = asRecord(model.input_state) ?? {};
  const paths = asArray(model.reasoning_paths).filter(isRecord);
  const selectedPathId = text(model.selected_path_id);
  const selectedPath = paths.find((path) => text(path.path_id) === selectedPathId);
  const steps = paths.flatMap((path) => asArray(path.steps).filter(isRecord));
  const target = asRecord(model.evaluation_target) ?? {};
  const pipeline = asRecord(runtime.pipeline_status) ?? {};
  const summary = asRecord(report.summary) ?? {};

  const vm: ReasoningRuntimeViewModel = {
    schemaVersion: SCHEMA,
    sourceRef: {
      sourceId: text(asRecord(runtime.source_ref)?.source_id) || "unavailable",
      sourceKind: text(asRecord(runtime.source_ref)?.source_kind) || "unknown",
      ...(asRecord(runtime.source_ref)?.source_path ? { sourcePath: text(asRecord(runtime.source_ref)?.source_path) } : {}),
    },
    runtimeStatus: {
      status: runtimeStatus(pipeline.status),
      runId: text(runtime.run_id),
      hasReasoningModel: Object.keys(model).length > 0,
      hasEvaluationReport: Object.keys(report).length > 0,
      diagnosticCount: diagnostics.length,
      fatalDiagnosticCount: diagnostics.filter((item) => item.severity === "fatal").length,
    },
    modelSummary: {
      available: Object.keys(model).length > 0,
      modelId: text(model.model_id) || "ReasoningModel unavailable",
      modelSchemaVersion: text(model.schema_version),
      selectedPathId,
      inputUnitCount: asArray(input.units).length,
      inputRelationCount: asArray(input.relations).length,
      reasoningPathCount: paths.length,
      reasoningStepCount: steps.length,
      knowledgeEmissionCount: asArray(model.knowledge_emissions).length,
      evaluationGoal: text(target.goal),
      requiredChecks: asArray(target.required_checks).map(text),
    },
    pipelineStatus: {
      status: runtimeStatus(pipeline.status),
      parserPassed: Boolean(pipeline.parser_passed),
      reasonIrAvailable: Boolean(pipeline.reason_ir_available),
      executionPlanAvailable: Boolean(pipeline.execution_plan_available),
      simulationAvailable: Boolean(pipeline.simulation_available),
      knowledgeAvailable: Boolean(pipeline.knowledge_available),
      diagnosticsCount: numberValue(pipeline.diagnostics_count),
    },
    inputState: {
      inputId: text(input.input_id),
      inputKind: text(input.input_kind),
      units: asArray(input.units).filter(isRecord).map((unit) => ({
        unitId: text(unit.unit_id),
        unitType: text(unit.unit_type),
        value: unit.value,
      })),
      relations: asArray(input.relations).filter(isRecord).map((relation) => ({
        relationId: text(relation.relation_id),
        relationType: text(relation.relation_type),
        source: text(relation.source),
        target: text(relation.target),
      })),
    },
    reasoningPath: {
      selectedPathId,
      selectedPathSignature: text(selectedPath?.path_signature),
      paths: paths.map((path) => ({
        pathId: text(path.path_id),
        pathSignature: text(path.path_signature),
        status: pathStatus(path.status),
        steps: asArray(path.steps).filter(isRecord).map((step) => ({
          stepId: text(step.step_id),
          stepType: text(step.step_type),
          source: text(step.source),
          operation: text(step.operation),
          target: text(step.target),
          evidenceRefs: asArray(step.evidence_refs).map(text),
        })),
      })),
    },
    knowledgeEmission: {
      emissions: asArray(model.knowledge_emissions).filter(isRecord).map((item) => ({
        knowledgeId: text(item.knowledge_id),
        sourceStepId: text(item.source_step_id),
        relation: text(item.relation),
        source: text(item.source),
        target: text(item.target),
        evidencePath: asArray(item.evidence_path).map(text),
        pathSignature: text(item.path_signature),
      })),
    },
    evaluationReport: {
      available: Object.keys(report).length > 0,
      reportId: text(report.report_id) || "EvaluationReport unavailable",
      status: evalStatus(summary.status),
      passed: Boolean(summary.passed),
      requiredChecksPassed: Boolean(summary.required_checks_passed),
      totalChecks: numberValue(summary.total_checks),
      passedChecks: numberValue(summary.passed_checks),
      failedChecks: numberValue(summary.failed_checks),
      warningChecks: numberValue(summary.warning_checks),
      fatalDiagnostics: numberValue(summary.fatal_diagnostics),
      checks: asArray(report.checks).filter(isRecord).map((check) => ({
        checkId: text(check.check_id),
        checkType: text(check.check_type),
        required: Boolean(check.required),
        status: checkStatus(check.status),
        passed: Boolean(check.passed),
        message: text(check.message),
        evidenceRefs: asArray(check.evidence_refs).map(text),
        details: asRecord(check.details) ?? {},
      })),
    },
    diagnostics: diagnosticsSummary(diagnostics),
    rawArtifacts: {
      runtimeResult: runtime,
      reasoningModel: model,
      evaluationReport: report,
    },
  };
  vm.rawArtifacts.viewModel = vm;
  return vm;
}

function collectDiagnostics(runtime: Record<string, unknown>, model: Record<string, unknown>, report: Record<string, unknown>): ReasoningDiagnosticItemViewModel[] {
  const items = [
    ...diagnosticsFrom("runtime", runtime.diagnostics),
    ...diagnosticsFrom("reasoning_model", model.diagnostics),
    ...diagnosticsFrom("evaluation_report", report.diagnostics),
  ];
  return items.sort((a, b) => {
    return (
      SEVERITY_ORDER[a.severity] - SEVERITY_ORDER[b.severity]
      || a.source.localeCompare(b.source)
      || a.code.localeCompare(b.code)
      || (a.location ?? "").localeCompare(b.location ?? "")
      || a.message.localeCompare(b.message)
    );
  });
}

function diagnosticsFrom(source: ReasoningDiagnosticItemViewModel["source"], value: unknown): ReasoningDiagnosticItemViewModel[] {
  return asArray(value).filter(isRecord).map((item) => ({
    source,
    code: text(item.code) || "RO-DIAG-UNKNOWN",
    severity: severity(item.severity),
    message: text(item.message),
    ...(item.location ? { location: text(item.location) } : {}),
  }));
}

function diagnosticsSummary(items: ReasoningDiagnosticItemViewModel[]) {
  return {
    total: items.length,
    fatal: items.filter((item) => item.severity === "fatal").length,
    error: items.filter((item) => item.severity === "error").length,
    warning: items.filter((item) => item.severity === "warning").length,
    info: items.filter((item) => item.severity === "info").length,
    items,
  };
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return value && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : null;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(asRecord(value));
}

function asArray(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

function text(value: unknown): string {
  return value == null ? "" : String(value);
}

function numberValue(value: unknown): number {
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

function runtimeStatus(value: unknown) {
  return value === "passed" || value === "failed" || value === "partial" || value === "fatal" ? value : "unknown";
}

function pathStatus(value: unknown) {
  return value === "selected" || value === "candidate" || value === "rejected" || value === "failed" ? value : "unknown";
}

function evalStatus(value: unknown) {
  return value === "passed" || value === "failed" || value === "warning" || value === "fatal" ? value : "unknown";
}

function checkStatus(value: unknown) {
  return value === "passed" || value === "failed" || value === "warning" || value === "skipped" || value === "fatal" ? value : "unknown";
}

function severity(value: unknown) {
  return value === "fatal" || value === "error" || value === "warning" || value === "info" ? value : "info";
}
